src/output.rs

```diff
@@ -321,56 +365,87 @@ where
-                        Some(&(Payload::Origin(origin), ref info)) => {
-                            (origin, info)
-                        },
-                        Some(_) => continue,
+                    let (origin, info) = match iter.next_with_info() {
+                        Some((origin, info)) => (origin, info),
+                        None => {
+                            self.formatter.intermission(target)?;
+                            break
+                        }
+                    };
+                    if let Some(selection) = self.selection.as_ref() {
+                        if !selection.include_origin(origin) {
+                            continue
+                        }
+                    }
+                    if *first {
+                        *first = false;
+                    }
+                    else {
+                        self.formatter.delimiter(target)?;
+                    }
+                    self.formatter.origin(origin, info, target)?;
+                    return Ok(true)
+                }
+                (
+                    true,
+                    StreamState::Key {
+                        iter: self.snapshot.clone().arc_router_keys_iter(),
+                        first: true
+                    }
+                )
+            }
+            StreamState::Key { ref mut iter, ref mut first } => {
+                loop {
+                    let (key, info) = match iter.next_with_info() {
+                        Some((key, info)) => (key, info),
                         None => {
                             self.formatter.footer(
                                 self.metrics.as_ref(), target
                             )?;
-                            self.state = StreamState::Done;
                             break
                         }
                     };
-                    index += 1;
                     if let Some(selection) = self.selection.as_ref() {
-                        if !selection.as_ref().include_origin(origin) {
+                        if !selection.include_router_key(key) {
                             continue
                         }
                     }
-                    if !first {
+                    if *first {
+                        *first = false;
+                    }
+                    else {
                         self.formatter.delimiter(target)?;
                     }
-                    self.formatter.origin(origin, info, target)?;
-                    self.state = StreamState::Origin { index, first: false };
-                    break
+                    self.formatter.router_key(key, info, target)?;
+                    return Ok(true)
                 }
-                Ok(true)
+                (true, StreamState::Done)
             }
-            StreamState::Done => Ok(false)
-        }
+            StreamState::Done => return Ok(false)
+        };
+        self.state = next;
+        Ok(res)
```

Comment:
`res` is always true here, so perhaps it can be removed from the `let (res, next) = match self.state` at line 370 and be simplified here to:
```suggestion
        Ok(true)
```

Reply:
Oh, good. That makes the code formatting much less ugly ;)